File: core/shared/src/main/scala/diplomacy/variant/standard/rule/RetreatOrderGenerator.scala
package diplomacy.variant.standard.rule

import diplomacy.rule.InvalidOrderMessage
import diplomacy.board.Power

class RetreatOrderGenerator[Turn_ <: Turn, Power_ <: Power] extends OrderGenerator[Turn_, Power_] {
  def ordersToSkipPhase(board: Board): Option[Set[Order]] = {
    val dislodgedUnits = board.unitStatuses collect {
      case (unit, x: UnitStatus.Dislodged[Power]) => unit -> x.attackedFrom
    }
    if (
      dislodgedUnits forall {
        case (unit, attackedFrom) =>
          StandardRuleUtils.locationsToRetreat(board)(unit, attackedFrom).isEmpty
      }
    ) {
      Option(
        (dislodgedUnits map { case (unit, _) => Order.Disband(unit) })(collection.breakOut)
      )
    } else {
      None
    }
  }
  def defaultOrder(board: Board)(unit: DiplomacyUnit): Order = Order.Disband(unit)
}

File: lib/variant/standard/rule/retreat-order-generator.js
const StandardRuleUtils = require('./standard-rule-utils')
const { Disband } = require('./order')

module.exports = class RetreatOrderGenerator {
  ordersToSkipPhase (board) {
    if (board.unitStatuses.size === 0) {
      return new Set()
    }
    if (
      [...board.unitStatuses].every(elem => {
        const [unit, status] = elem
        const locations = StandardRuleUtils.locationsToRetreat(board, unit, status.attackedFrom)
        return locations.size === 0
      })
    ) {
      // Disband all dislodged units
      const retval = new Set()
      board.unitStatuses.forEach((status, unit) => {
        retval.add(new Disband(unit))
      })
      return retval
    }
    return null
  }
  defaultOrderOf (board, unit) {
    return new Disband(unit)
  }
}
